src/main/cpp/apple_fsnotifier.cpp

```diff
@@ -0,0 +1,179 @@
+
+/*
+ * Apple specific functions.
+ */
+#if defined(__APPLE__)
+
+#include "native.h"
+#include "generic.h"
+#include <CoreServices/CoreServices.h>
+#include <pthread.h>
+#include <strings.h>
+
+CFMutableArrayRef rootsToWatch = NULL;
+FSEventStreamRef watcherStream = NULL;
+pthread_t watcherThread = NULL;
+// store the callback object, as we need to invoke it once file change is detected.
+jobject watcherCallback = NULL;
+JavaVM* jvm = NULL;
+CFRunLoopRef threadLoop = NULL;
+bool invalidStateDetected = false;
+
+static void reportEvent(const char *event, char *path) {
+    size_t len = 0;
+    if (path != NULL) {
+        len = strlen(path);
+        for (char *p = path; *p != '\0'; p++) {
+            if (*p == '\n') {
+                *p = '\0';
+            }
+        }
+    }
+    JNIEnv* env;
+    int getEnvStat = jvm->GetEnv((void **)&env, JNI_VERSION_1_6);
+    if (getEnvStat == JNI_EDETACHED) {
+        if (jvm->AttachCurrentThread((void **) &env, NULL) != JNI_OK) {
+            invalidStateDetected = true;
+            return;
+        }
+    } else if (getEnvStat == JNI_EVERSION) {
+        invalidStateDetected = true;
+        return;
+    }
+
+    jclass callback_class = env->GetObjectClass(watcherCallback);
+    jmethodID methodCallback = env->GetMethodID(callback_class, "pathChanged", "(Ljava/lang/String;)V");
+    env->CallVoidMethod(watcherCallback, methodCallback, env->NewStringUTF(path));
+}
+
+static void callback(ConstFSEventStreamRef streamRef,
+                     void *clientCallBackInfo,
+                     size_t numEvents,
+                     void *eventPaths,
+                     const FSEventStreamEventFlags eventFlags[],
+                     const FSEventStreamEventId eventIds[]) {
+    if (invalidStateDetected) return;
+    char **paths = (char**) eventPaths;
+
+    for (int i = 0; i < numEvents; i++) {
+        // TODO[max] Lion has much more detailed flags we need accurately process. For now just reduce to SL events range.
+        FSEventStreamEventFlags flags = eventFlags[i] & 0xFF;
+        if ((flags & kFSEventStreamEventFlagMustScanSubDirs) != 0) {
+            reportEvent("RECDIRTY", paths[i]);
+        } else if (flags != kFSEventStreamEventFlagNone) {
+            reportEvent("RESET", NULL);
+        } else {
+            reportEvent("DIRTY", paths[i]);
+        }
+    }
+}
+
+JNIEXPORT void JNICALL
+Java_net_rubygrapefruit_platform_internal_jni_OsxFileEventFunctions_createWatch(JNIEnv *env, jclass target, jstring path, jobject result) {
+    if (rootsToWatch == NULL) {
+        invalidStateDetected = false;
+        rootsToWatch = CFArrayCreateMutable(NULL, 0, NULL);
+        if (rootsToWatch == NULL) {
+            mark_failed_with_errno(env, "Could not allocate array to store roots to watch.", result);
+            return;
+        }
+    }
+    char* pathString = java_to_char(env, path, result);
+    if (pathString == NULL) {
+        return;
```

Comment:
I'd like to revamp our `null` handling completely when we move to the even- driven API. Let's keep it this way for now as it will be refactored later.